1. Mario García
– 2. Ricardo Ibarra 
– 3. Paula Pelayo

# PROYECTO MÓDULO 2

### Objetivos
#### General:
Mediante el uso de programación lineal, optimizar y maximizar las ganancias de una inversión.

#### Específico:
Utilizar teoría de programación lineal, y eventualmente herramientas computacionales, para llevar a cabo un modelo de optimización en base a un portafolio de inversión, el cual necesita ser optimizado tomando en cuenta los requisitos del cliente.


### Problema:

Hay un inversionista que quiere invertir en la Bolsa de Valores en seis diferentes empresas.

Cuenta con un capital máximo de $50,000.00 (cincuenta mil dólares USD) y quiere dividir su inversión en las siguientes empresas:

- Tesla, Inc.
- Amazon.com, Inc.
- Walmart Inc.
- Apple Inc.
- Johnson & Johnson 
- NVIDIA Corporation

El inversionista desea tener un número de acciones definidas en las siguientes empresas:

- Tesla, Inc. (40 acciones)
- Apple Inc. (50 acciones)

Además de eso, quiere tener un mínimo de acciones en las siguientes empresas:

- Walmart Inc. (30 acciones)
- Johnson & Johnson  (20 acciones)

Y un máximo de acciones en la siguiente empresa:

- Amazon.com, Inc. (80 acciones)

Por último, desea tener un mínimo de acciones en diferentes sectores del mercado bursátil:

- Tecnología electrónica (70 acciones)
- Comercio minorista (50 acciones)

El objetivo será maximizar el valor de su portafolio en base a lo que quiere invertir.

**El precio de las acciones está representado en la siguiente tabla, la cual fue recuperado de Yahoo Finance:**

Empresa|Precio de acción|Sector
:----|----|----
Tesla|$\$$197.58|Servicios y bienes de consumo no básico
Amazon|$\$$100.61|Comercio minorista
Walmart|$\$$140.42|Comercio minorista
Apple|$\$$159.28|Tecnología electrónica
Johnson & Johnson|$\$$153.89|Tecnología de salud
NVIDIA|$\$$261.99|Tecnología electrónica

Dentro del mercado bursátil, se utilizan herramientas llamadas múltiplos bursátiles, los cuales se basan en las tendencias de las acciones para brindar información útil acerca de los posibles sucesos de las acciones en el futuro. Esto debido a que no hay una forma exacta de saber la ganancia esperada de una acción.

Una de estas herramientas es el _precio utilidad por acción (P/U)_ el cual se obtiene de dividir el market cap (valor de la empresa) entre la utilidad anual de la empresa. Esta herramienta señala los años aproximados en que tardarías en recuperar el 100% de tu inversión.

$$P/U = \frac{Market cap}{Utilidades}$$

A su vez, el recíproco de este representa, en vía utilidades, cuánto se generó en rendimiento. Por lo tanto, se puede tomar en cuenta esta tasa de rendimiento para predecir la ganancia que se espera de cada acción

$$\text{Tasa de rendimiento} = \frac{1}{P/U}$$

Por lo tanto, en la siguiente tabla se muestran las tasas de rendimiento calculadas de cada acción, el market cap fue recuperado de yahoo finance:

Empresa|Market Cap|Utilidad|Precio utilidad por acción|Tasa de rendimiento
:----|----|----|----|----
Tesla|$\$$623,802,000,000.00 |$\$$13,832,000,000.00|45.098467|2.2174%
Amazon|$\$$1,028,000,000,000.00|$\$$12,248,000,000.00|83.93207|1.1914%
Walmart|$\$$379,091,000,000.00|$\$$20,428,000,000.00|18.55742|5.3887%
Apple|$\$$2,538,000,000,000.00|$\$$113,485,000,000.00|22.36419|4.4714%
Johnson & Johnson|$\$$477,242,000,000.00|$\$$23,703,000,000.00|20.13424|4.9667%
NVIDIA|$\$$675,282,000,000.00|$\$$5,577,000,000.00|121.08338|0.8259%

### Modelo que representa el problema

**1. Criterio de optimización:**

- Maximizar

**2. Función:** 

La función a optimizar es el monto invertido en acciones de cada empresa multiplicado por la tasa de rendimiento anual. Se desea maximizar dicha función para llegar a la máxima ganancia posible.

$$max. 0.022x_1 + 0.011x_2 + 0.053x_3 + 0.044x_4 + 0.049x_5 + 0.0082x_6$$


**3. Definir Variables:**

Cada x representa el monto invertido en acciones de las siguientes empresas:

- $x_1$: Tesla, Inc.
- $x_2$: Amazon.com, Inc.
- $x_3$: Walmart Inc.
- $x_4$: Apple Inc.
- $x_5$: Johnson & Johnson 
- $x_6$: NVIDIA Corporation

**4. Restricciones:**

El monto total de todas las acciones no debe superar los \$50,000
- $x_{1} + x_{2} + x_{3} + x_{4} + x_{5} + x_{6}\leq 50000$

El número de acciones (monto entre precio de acción) de ciertas empresas (Tesla y Apple) deben ser un valor único, por lo tanto:
- $\frac{x_{1}}{197.58} = 40$
- $\frac{x_{4}}{159.28} = 50$

El número de acciones de ciertas empresas (Walmart y Johnson & Johnson) debe ser como mínimo: 
- $\frac{x_{3}}{140.42} \geq 30$
- $\frac{x_{5}}{153.89} \geq 20$

El número de acciones de Amazon no debe de superar un número en específico:
- $\frac{x_{2}}{100.61} \leq 80$

El número de acciones en el sector de Tecnología Electrónica (Apple y NVIDIA) debe ser como mínimo 70
- $(\frac{x_{4}}{159.28}+\frac{x_{6}}{261.99}) \geq 70$

El número de acciones en el sector de Comercio minorista (Amazon y Walmart) debe ser como mínimo 50
- $(\frac{x_{2}}{100.61}+\frac{x_{3}}{140.42}) \geq 50$




5. Matrices
\begin{gather}
x
 =
  \begin{bmatrix}
   x_1 \\
   x_2 \\
   x_3 \\
   x_4 \\
   x_5 \\
   x_6 \\
   \end{bmatrix}
\end{gather}

\begin{gather}
c
 =
  \begin{bmatrix}
   0.022 \\
   0.011 \\
   0.053 \\
   0.044 \\
   0.049 \\
   0.0082 \\
   \end{bmatrix}
\end{gather}

\begin{gather}
A_{ub}
 =
  \begin{bmatrix}
   1 & 1 & 1 & 1 & 1 & 1\\
   0 & 0 & \frac{-1}{140.42} & 0 & 0 & 0\\
   0 & 0 & 0 & 0 & \frac{-1}{153.89} & 0\\
   0 & \frac{1}{100.61} & 0 & 0 & 0 & 0\\
   0 & 0 & 0 & \frac{-1}{159.28} & 0 & \frac{-1}{261.99}\\
   0 & \frac{-1}{100.61} & \frac{-1}{140.42} & 0 & 0 & 0\\
   \end{bmatrix}
\end{gather}

\begin{gather}
b_{ub}
 =
  \begin{bmatrix}
   50000\\
   -30\\
   -20\\
   80\\
   -70\\
   -50\\
   \end{bmatrix}
\end{gather}

\begin{gather}
A_{eq}
 =
  \begin{bmatrix}
   \frac{1}{197.58} & 0 & 0 & 0 & 0 & 0\\
   0 & 0 & 0 & \frac{1}{159.28} & 0 & 0\\
   \end{bmatrix}
\end{gather}

\begin{gather}
b_{eq}
 =
  \begin{bmatrix}
   40\\
   50\\
   \end{bmatrix}
\end{gather}


### Solución del problema de optimización

In [3]:
import numpy as np
import scipy.optimize as opt

In [10]:
empresas = ["Tesla", "Amazon", "Walmart", "Apple", "Johnson & Johnson", "NVIDIA"]
c = np.array([-0.022, -0.011, -0.053, -0.044, -0.049, -0.0082])
A_ub = np.array([[1, 1, 1, 1, 1, 1],
                [0, 0, -1/140.42, 0, 0, 0],
                [0, 0, 0, 0, -1/153.89, 0],
                [0, 1/100.61, 0, 0, 0, 0],
                [0, 0, 0, -1/159.28, 0, -1/261.99],
                [0, -1/100.61, -1/140.42, 0, 0, 0]])
b_ub = np.array([50000, -30, -20, 80, -70, -50])
A_eq = np.array([[1/197.58, 0, 0, 0, 0, 0],
                [0, 0, 0, 1/159.28, 0, 0]])
b_eq = np.array([40, 50])
sol = opt.linprog(c, A_ub, b_ub, A_eq, b_eq)
print("Ganancia anual esperada: $",round(sol.get("fun"), 2)*-1)
print("")
for i in range(6):
    print("$", round(sol.get("x")[i]), "invertido en", empresas[i])
    

Ganancia anual esperada: $ 2086.27

$ 7903 invertido en Tesla
$ 0 invertido en Amazon
$ 25815 invertido en Walmart
$ 7964 invertido en Apple
$ 3078 invertido en Johnson & Johnson
$ 5240 invertido en NVIDIA


Por lo tanto, el número de acciones a invertir en cada empresa son los siguientes:

- Tesla: 40 acciones
- Amazon: 0 acciones
- Walmart: 183 acciones
- Apple: 50 acciones
- Johnson & Johnson: 20 acciones
- NVIDIA: 20 acciones

El algoritmo de optimización se condujo a una solución factible debido a que, los resultados obtenidos representan el monto a invertir en cada empresa respetando las cláusulas presentadas por el inversionista, y al mismo tiempo, maximizando su ganancia anual esperada.

Empresa|Esperado (acciones)|Obtenido (acciones)
:----|----|----
Tesla|40|40
Amazon|Máximo 80|0
Walmart|Mínimo 30|183
Apple|50|50
Johnson & Johnson|Mínimo 20|20 
NVIDIA| - |20

Sector|Esperado (acciones)|Obtenido (acciones)
:----|----|----
Tecnología Electrónica|Mínimo 70 |70
Comercio minorista|Mínimo 50|183

**Resultados:**

Empresa|Acciones|Precio|Monto|Tasa rendimiento|Ganancia anual
:----|----|----|----|----|----
Tesla|40|$\$$197.58|$\$$7,903|2.21%|$\$$156.04
Amazon|0|$\$$100.61|$\$$0|1.19%|$\$$0
Walmart|183|$\$$140.42|$\$$25,815|5.38%|$\$$1,368.19
Apple|50|$\$$159.28|$\$$7,964|4.47%|$\$$350.41
Johnson & Johnson|20|$\$$153.89|$\$$3,078|4.96%|$\$$150.82
NVIDIA|20|$\$$261.99|$\$$5,240|0.82%|$\$$42.96
|||||$\$$2,086.44

Como conclusión, se logró maximizar la ganancia del inversionista, esto con base en las restricciones que él mismo propuso en un primer momento. Tanto las empresas en las que obligatoriamente quería invertir, como el mínimo y máximo de acciones en distintos sectores. De esta forma, se obtiene un portafolio de inversión diverso, con diferentes empresas y la máxima ganancia posible.

Sin embargo, consideramos que a pesar de que se maximizó la ganancia esperada con las cláusulas dadas por el cliente, no hay necesidad de invertir un número tan específico de acciones en las empresas que él deseaba, de esta forma, hay otras maneras de invertir para conseguir un portafolio diverso y tal vez con más ganancias.

También hay que tener en cuenta que los múltiplos bursátiles, en este caso el usado Precio Utilidad, están basados en datos e información del pasado, por lo que no pueden ser tan acertados para predecir el futuro de las acciones y sus valores. Debido a esto, la tasa de rendimiento usada es calculada a partir de múltiplos bursátiles, por lo cual su uso dentro del ejercicio es una predicción de la ganancia, la cual puede variar demasiado durante el año de vida de la acción. Además en el mundo bursátil, las acciones siempre cambian debido al comportamiento humano, por lo tanto pueden llegar a ser impredecibles.

**Referencias:**

- Yahoo Finance. (s. f.). Recuperado 21 de marzo de 2023, de https://finance.yahoo.com/